[Tan Lip Wei] iP

lipwei1808 opened the pull request:
### Duke Chatbot

> "Naming and remembering things is hard." - Me

Duke is ~~dumb~~ smart and can
- Track your tasks
- Delete tasks
- Mark them as completed
- Find tasks

#### Getting started

1. Head over [here](https://github.com/lipwei1808/ip) 
2. Clone the project
     - `git clone <link>`
3. Run it locally

And there your have it, **DUKE at your service :+1:**

#### Trying it out

- [ ] Add a task
- [ ] Mark a task
- [ ] Delete a task
- [ ] Find a task

#### For programmers

If you would like to edit the file, start from the entry point (main method) in Duke.java

```
public static void main(String[] args) {
        new Duke().run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lipwei1808 committed:
Level-0

lipwei1808 committed:
Level-1

lipwei1808 committed:
Level-2

lipwei1808 committed:
Level-3

lipwei1808 committed:
Level-4

lipwei1808 committed:
Fix bug with event and added testing

lipwei1808 committed:
Added exception handling for adding of tasks

lipwei1808 committed:
Added delete functionality and updated test

lipwei1808 committed:
Change bot name ah

lipwei1808 committed:
Add saving functionality on new task

lipwei1808 committed:
Add auto-save mark and unmark information

lipwei1808 committed:
Merge branch 'branch-Level-7'

lipwei1808 committed:
Add DateTime parsing by the chatbot

lipwei1808 committed:
Allow for formatted dates when listing tasks

lipwei1808 committed:
Merge branch 'branch-Level-8'

lipwei1808 committed:
Test cases

lipwei1808 committed:
Add UI, Parser, Command and Storage class

lipwei1808 committed:
cleanup code

lipwei1808 committed:
Merge branch 'branch-more-oop'

lipwei1808 committed:
Refactor to packages

lipwei1808 committed:
Merge branch 'branch-packages'

lipwei1808 committed:
Update class used in build gradle

lipwei1808 committed:
Merge branch 'add-gradle-support'

lipwei1808 committed:
Add junit test cases

lipwei1808 committed:
Merge branch 'branch-junit'

lipwei1808 committed:
Add handling for missing directories when creating data file

lipwei1808 committed:
Merge branch 'branch-jar'

lipwei1808 committed:
Add FindCommand and Parser to handle new case

lipwei1808 committed:
Merge branch 'branch-Level-9'

lipwei1808 committed:
Refactor lines printing in UI class

lipwei1808 committed:
Refactor codebase to align with Coding Standards

lipwei1808 committed:
Fix lines that are too long

lipwei1808 committed:
Merge branch 'branch-A-CodingStandard'

lipwei1808 committed:
Add javadocs

lipwei1808 committed:
Complete javadocs for rest of the classes

lipwei1808 committed:
Write javaDocs for new methods

lipwei1808 committed:
Merge branch 'branch-A-JavaDoc'

lipwei1808 committed:
Fix style

lipwei1808 committed:
Fix style for tests

lipwei1808 committed:
Merge branch 'branch-A-CheckStyle'

lipwei1808 committed:
Add javafx and refactor UI class

lipwei1808 committed:
Add theme color and fix layout issues

lipwei1808 committed:
Merge branch 'branch-Level-10'

lipwei1808 committed:
Add javadocs for ui classes

TeeRenJing commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+ * A task object that tracks Deadlines.
+ */
+public class Deadline extends Task {
+    private final LocalDateTime by;
```
Perhaps capitalize constant names?

lipwei1808 replied:
I think this is not really a constant in this case as if we allow for future edits to tasks, then it wouldn't really be a constant right

TeeRenJing commented on src/main/java/duke/ui/DialogBox.java:
```diff
@@ -0,0 +1,83 @@
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
```
I like how imported classes are listed explicitly!

TeeRenJing commented on src/main/java/duke/tasks/Tasks.java:
```diff
@@ -0,0 +1,72 @@
+ * Represents a collection of Task instances.
+ */
+public class Tasks {
+    private final List<Task> tasks = new ArrayList<>();
```
I like how plural form is used on names representing a collection of objects.

TeeRenJing commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,122 @@
+     * @param dateTime the LocalDateTime object to be stringified
+     * @return text string of the LocalDateTime object
+     */
+    public String stringifyDate(LocalDateTime dateTime) {
```
I like how method names are verbs and written in camelCase! 👍

TeeRenJing reviewed (commented):
Looks good to me! 👍 
Just a few nits to fix.

lunaflight commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return a boolean value if the program should exit
+     */
+    public boolean isExit() {
```
Do you think this is an non-intuitive use of a function name?

It seems that this is only ever used to check if it is an Exit command - could using `instanceof ExitCommand` be cleaner?

lipwei1808 replied:
That's true, I think it was added as in the future if further commands were added that could lead to the function exiting, they could easily override this function.

I agree with the naming of the function, will update it to be more intuitive

lunaflight commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public String execute(Tasks tasks, Ui ui, Storage storage, boolean isRestoring) {
+        List<Task> res = new ArrayList<>();
```
What do you think of a more descriptive variable name, such as `foundTasks`?

lipwei1808 replied:
That's a good idea

lunaflight commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     */
+    protected static String[] parseText(String text) {
+        String[] words = text.trim().split(" ");
+        String[] remaining = Arrays.copyOfRange(words, 1, words.length);
```
What do you think of a variable name with a plural s, such as `remainingWords`?

lipwei1808 replied:
Thanks for the suggestion

lunaflight commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,136 @@
+     * @param isMark boolean value whether to mark the task
+     * @param task the task to be marked
+     */
+    public String showSuccessMark(boolean isMark, Task task) {
```
You may want to rename the function name to something more intuitive, such as `getSuccessMarkMessage`.
`show` might give off the impression that it is a `void` function.

I noticed this in other function names in this class as well.

lunaflight reviewed (commented):
Looks good, just a few nits :)

lipwei1808 committed:
fix the nits

lipwei1808 committed:
Add assertions for important assumptions in program

There are some important assumptions made in the program made in
the program that should hold. For example, arguments should not
be an empty string, etc.

This allows for early catching of bugs in the program.

Java Assertions can be used to assert a certain value in the program
at a certain point in the code. Hence, adding assertions to assert
a certain value can target this problem.

lipwei1808 committed:
Improve code quality by refactoring methods

Long method with too many lines could cause future developers reading
the code from quickly understanding what the method is doing. It could
also be a sign that the method is not following the SLAP principle.

There could also be high level of nestings leading to the arrowhead
style code - making it harder for people to read the code.

By extracting certain parts of the method into their own method can
reduce the number of levels of abstraction in a single method.

lipwei1808 committed:
Merge pull request #1 from lipwei1808/branch-A-Assertions

Add assertions for important assumptions in program

lipwei1808 committed:
Merge pull request #2 from lipwei1808/branch-A-CodeQuality

Improve code quality by refactoring methods

lipwei1808 committed:
Add a CI file

lipwei1808 committed:
Update test case for Parser

lipwei1808 committed:
Merge pull request #3 from lipwei1808/branch-CI

Add a CI file

lipwei1808 committed:
Chcek for duplicate tasks

lipwei1808 committed:
Merge pull request #4 from lipwei1808/branch-BCD

Chcek for duplicate tasks

lipwei1808 committed:
make duke auto size

lipwei1808 committed:
Improve the UI for error and dialog box

lipwei1808 committed:
Merge branch 'branch-A-BetterGui'

lipwei1808 committed:
Update README and add UI pic

lipwei1808 committed:
Merge branch 'branch-UserGuide'

lipwei1808 committed:
Update UserGuide on root README

lipwei1808 committed:
Update some javadocs

lipwei1808 committed:
Merge branch 'branch-CodeQuality'

lipwei1808 committed:
update build gradle
